In [142]:
import pandas as pd 
import numpy as np  
import matplotlib.pyplot as plt  
import seaborn as sns

In [143]:
%matplotlib inline

In [144]:
gdp_df=pd.read_csv('data/gdp_percapita.csv',nrows=6729) #imports data w/o footnotes - does not affect data table in data folder

In [145]:
print(gdp_df.head(6))

  Country or Area  Year        Value  Value Footnotes
0     Afghanistan  2018  1734.723214              NaN
1     Afghanistan  2017  1758.465636              NaN
2     Afghanistan  2016  1757.023490              NaN
3     Afghanistan  2015  1766.593077              NaN
4     Afghanistan  2014  1795.735834              NaN
5     Afghanistan  2013  1807.762344              NaN


In [146]:
print(gdp_df.tail(6))

     Country or Area  Year        Value  Value Footnotes
6723        Zimbabwe  1995  2736.486436              NaN
6724        Zimbabwe  1994  2768.309953              NaN
6725        Zimbabwe  1993  2572.870395              NaN
6726        Zimbabwe  1992  2591.007534              NaN
6727        Zimbabwe  1991  2906.272849              NaN
6728        Zimbabwe  1990  2819.549467              NaN


In [147]:
internet_df=pd.read_csv('data/internet_use.csv',nrows=4495)#imports data w/o footnotes - does not affect data table in data folder

In [148]:
print(internet_df.head(6))

  Country or Area  Year     Value  Value Footnotes
0     Afghanistan  2014  6.390000              NaN
1     Afghanistan  2013  5.900000              NaN
2     Afghanistan  2012  5.454545              NaN
3     Afghanistan  2011  5.000000              NaN
4     Afghanistan  2010  4.000000              NaN
5     Afghanistan  2009  3.550000              NaN


In [149]:
print(internet_df.tail(6))

     Country or Area  Year     Value  Value Footnotes
4489        Zimbabwe  1998  0.081648              NaN
4490        Zimbabwe  1997  0.033080              NaN
4491        Zimbabwe  1996  0.016790              NaN
4492        Zimbabwe  1995  0.007684              NaN
4493        Zimbabwe  1994  0.001739              NaN
4494        Zimbabwe  1990  0.000000              NaN


In [150]:
#np_gdp=np.array(gdp_df) -thought I had to convert to numpy array to see shape
print(gdp_df.shape)

(6729, 4)


In [151]:
#np_internet=np.array(internet_df) -thought I had to convert to numpy array to see shape
print(internet_df.shape)

(4495, 4)


In [152]:
#print(type(np_gdp))
print(type(gdp_df))

<class 'pandas.core.frame.DataFrame'>


In [153]:
#print(np.info(np_pgd))
print(gdp_df.info)

<bound method DataFrame.info of      Country or Area  Year        Value  Value Footnotes
0        Afghanistan  2018  1734.723214              NaN
1        Afghanistan  2017  1758.465636              NaN
2        Afghanistan  2016  1757.023490              NaN
3        Afghanistan  2015  1766.593077              NaN
4        Afghanistan  2014  1795.735834              NaN
...              ...   ...          ...              ...
6724        Zimbabwe  1994  2768.309953              NaN
6725        Zimbabwe  1993  2572.870395              NaN
6726        Zimbabwe  1992  2591.007534              NaN
6727        Zimbabwe  1991  2906.272849              NaN
6728        Zimbabwe  1990  2819.549467              NaN

[6729 rows x 4 columns]>


In [154]:
print(gdp_df.dtypes)



Country or Area     object
Year                 int64
Value              float64
Value Footnotes    float64
dtype: object


In [155]:
print(internet_df.dtypes)

Country or Area     object
Year                 int64
Value              float64
Value Footnotes    float64
dtype: object


In [156]:
print(gdp_df.tail(10))

     Country or Area  Year        Value  Value Footnotes
6719        Zimbabwe  1999  3054.064189              NaN
6720        Zimbabwe  1998  3099.076182              NaN
6721        Zimbabwe  1997  3036.422224              NaN
6722        Zimbabwe  1996  2985.856605              NaN
6723        Zimbabwe  1995  2736.486436              NaN
6724        Zimbabwe  1994  2768.309953              NaN
6725        Zimbabwe  1993  2572.870395              NaN
6726        Zimbabwe  1992  2591.007534              NaN
6727        Zimbabwe  1991  2906.272849              NaN
6728        Zimbabwe  1990  2819.549467              NaN


In [157]:
print(internet_df.tail(10))

     Country or Area  Year     Value  Value Footnotes
4485        Zimbabwe  2002  3.994356              NaN
4486        Zimbabwe  2001  0.799846              NaN
4487        Zimbabwe  2000  0.401434              NaN
4488        Zimbabwe  1999  0.161676              NaN
4489        Zimbabwe  1998  0.081648              NaN
4490        Zimbabwe  1997  0.033080              NaN
4491        Zimbabwe  1996  0.016790              NaN
4492        Zimbabwe  1995  0.007684              NaN
4493        Zimbabwe  1994  0.001739              NaN
4494        Zimbabwe  1990  0.000000              NaN


In [158]:
gdp_df.pop('Value Footnotes')

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
        ..
6724   NaN
6725   NaN
6726   NaN
6727   NaN
6728   NaN
Name: Value Footnotes, Length: 6729, dtype: float64

In [159]:
print(gdp_df)

     Country or Area  Year        Value
0        Afghanistan  2018  1734.723214
1        Afghanistan  2017  1758.465636
2        Afghanistan  2016  1757.023490
3        Afghanistan  2015  1766.593077
4        Afghanistan  2014  1795.735834
...              ...   ...          ...
6724        Zimbabwe  1994  2768.309953
6725        Zimbabwe  1993  2572.870395
6726        Zimbabwe  1992  2591.007534
6727        Zimbabwe  1991  2906.272849
6728        Zimbabwe  1990  2819.549467

[6729 rows x 3 columns]


In [160]:
internet_df.pop('Value Footnotes')

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
        ..
4490   NaN
4491   NaN
4492   NaN
4493   NaN
4494   NaN
Name: Value Footnotes, Length: 4495, dtype: float64

In [161]:
print(internet_df)

     Country or Area  Year     Value
0        Afghanistan  2014  6.390000
1        Afghanistan  2013  5.900000
2        Afghanistan  2012  5.454545
3        Afghanistan  2011  5.000000
4        Afghanistan  2010  4.000000
...              ...   ...       ...
4490        Zimbabwe  1997  0.033080
4491        Zimbabwe  1996  0.016790
4492        Zimbabwe  1995  0.007684
4493        Zimbabwe  1994  0.001739
4494        Zimbabwe  1990  0.000000

[4495 rows x 3 columns]


In [162]:
gdp_df_new=gdp_df.rename(columns={'Country or Area':'Country','Value':'GDP_Per_Capita'})
print(gdp_df_new.head(6))

       Country  Year  GDP_Per_Capita
0  Afghanistan  2018     1734.723214
1  Afghanistan  2017     1758.465636
2  Afghanistan  2016     1757.023490
3  Afghanistan  2015     1766.593077
4  Afghanistan  2014     1795.735834
5  Afghanistan  2013     1807.762344


In [163]:
print(gdp_df_new.tail(6))

       Country  Year  GDP_Per_Capita
6723  Zimbabwe  1995     2736.486436
6724  Zimbabwe  1994     2768.309953
6725  Zimbabwe  1993     2572.870395
6726  Zimbabwe  1992     2591.007534
6727  Zimbabwe  1991     2906.272849
6728  Zimbabwe  1990     2819.549467


In [164]:
internet_df_new=internet_df.rename(columns={'Country or Area':'Country','Value':'Internet_Users_Pct.'})
print(internet_df_new.head(6))

       Country  Year  Internet_Users_Pct.
0  Afghanistan  2014             6.390000
1  Afghanistan  2013             5.900000
2  Afghanistan  2012             5.454545
3  Afghanistan  2011             5.000000
4  Afghanistan  2010             4.000000
5  Afghanistan  2009             3.550000


In [172]:
print(internet_df_new.tail(6))

       Country  Year  Internet_Users_Pct.
4489  Zimbabwe  1998             0.081648
4490  Zimbabwe  1997             0.033080
4491  Zimbabwe  1996             0.016790
4492  Zimbabwe  1995             0.007684
4493  Zimbabwe  1994             0.001739
4494  Zimbabwe  1990             0.000000


In [184]:
gdp_and_internet_use=pd.merge(gdp_df_new,internet_df_new,on=['Country','Year'],how='inner')
print(gdp_and_internet_use)

          Country  Year  GDP_Per_Capita  Internet_Users_Pct.
0     Afghanistan  2014     1795.735834             6.390000
1     Afghanistan  2013     1807.762344             5.900000
2     Afghanistan  2012     1772.764974             5.454545
3     Afghanistan  2011     1626.764793             5.000000
4     Afghanistan  2010     1671.581238             4.000000
...           ...   ...             ...                  ...
3514     Zimbabwe  1997     3036.422224             0.033080
3515     Zimbabwe  1996     2985.856605             0.016790
3516     Zimbabwe  1995     2736.486436             0.007684
3517     Zimbabwe  1994     2768.309953             0.001739
3518     Zimbabwe  1990     2819.549467             0.000000

[3519 rows x 4 columns]


An outer join creates a dataframe with 7705 rows, a left join creates a dataframe with 6729 rows, and a right join createsa dataframe with 4495 rows. All three of these create null values, which the inner join avoids. Since there needs to be values for all four columns in every row in order to answer some of the questions, an inner join is preferable. 

gdp_df has 6729 rows (after dropping footnotes) and internet_df has 4495 rows (after dropping footnotes) 
gdp_and_internet_use has 3519 rows: 3210 fewer than gdp_df and 1285 fewer than internet_df

In [190]:
gdp_internet_subset=gdp_and_internet_use[gdp_and_internet_use['Year'].isin(['2004','2009','2014'])]
print(gdp_internet_subset)

          Country  Year  GDP_Per_Capita  Internet_Users_Pct.
0     Afghanistan  2014     1795.735834             6.390000
5     Afghanistan  2009     1502.354073             3.550000
10    Afghanistan  2004     1025.208245             0.105809
13        Albania  2014    10700.993216            60.100000
18        Albania  2009     9525.377979            41.200000
...           ...   ...             ...                  ...
3480       Zambia  2009     3115.937579             6.310000
3485       Zambia  2004     2415.590117             2.013550
3497     Zimbabwe  2014     2506.490220            19.890000
3502     Zimbabwe  2009     1632.999866            11.360000
3507     Zimbabwe  2004     2104.143512             6.564045

[488 rows x 4 columns]


In [192]:
gdp_df_subset=gdp_df[gdp_df['Year'].isin(['2004','2009','2014'])]
print(gdp_df_subset)

     Country or Area  Year         Value
4        Afghanistan  2014   1795.735834
9        Afghanistan  2009   1502.354073
14       Afghanistan  2004   1025.208245
21           Albania  2014  10700.993216
26           Albania  2009   9525.377979
...              ...   ...           ...
6680          Zambia  2009   3115.937579
6685          Zambia  2004   2415.590117
6704        Zimbabwe  2014   2506.490220
6709        Zimbabwe  2009   1632.999866
6714        Zimbabwe  2004   2104.143512

[719 rows x 3 columns]


In [193]:
internet_df_subset=internet_df[gdp_df['Year'].isin(['2004','2009','2014'])]
print(internet_df_subset)

     Country or Area  Year      Value
4        Afghanistan  2010   4.000000
9        Afghanistan  2005   1.224148
14       Afghanistan  1990   0.000000
21           Albania  2008  23.860000
26           Albania  2003   0.971900
...              ...   ...        ...
4448           Yemen  1997   0.015025
4453          Zambia  2012  13.468200
4472          Zambia  1990   0.000000
4477        Zimbabwe  2010  11.500000
4482        Zimbabwe  2005   8.015978

[477 rows x 3 columns]


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.
